src/Core/Python/PythonDatatypeConverter.cc

```diff
@@ -451,50 +476,58 @@ namespace
   }
 }
 
-Variable SCIRun::Core::Python::convertPythonObjectToVariable(const boost::python::object& object)
+const std::string SCIRun::Core::Python::getClassName(const py::object& object)
+{
+  py::extract<py::object> extractor(object);
+  return py::extract<std::string>(extractor().attr("__class__").attr("__name__"));
+}
+
+Variable SCIRun::Core::Python::convertPythonObjectToVariable(const py::object& object)
 {
   /// @todo: yucky
+
+  auto classname = getClassName(object);
+  if (classname == "int")
   {
-    boost::python::extract<int> e(object);
-    if (e.check())
-    {
-      return makeVariable("int", e());
-    }
+    py::extract<int> e(object);
+    return makeVariable(classname, e());
   }
+  else if (classname == "float")
   {
-    boost::python::extract<double> e(object);
-    if (e.check())
-    {
-      return makeVariable("double", e());
-    }
+    py::extract<float> e(object);
+    return makeVariable(classname, e());
   }
+  else if (classname == "double")
   {
-    boost::python::extract<std::string> e(object);
-    if (e.check())
-    {
-      return makeVariable("string", e());
-    }
+    py::extract<double> e(object);
+    return makeVariable(classname, e());
   }
+  else if (classname == "str")
   {
-    boost::python::extract<bool> e(object);
-    if (e.check())
-    {
-      return makeVariable("bool", e());
-    }
+    py::extract<std::string> e(object);
+    return makeVariable(classname, e());
+  }
+  else if (classname == "bool")
+  {
+    py::extract<bool> e(object);
+    return makeVariable(classname, e());
   }
+  else if (classname == "list")
+  {
+    Variable::List vars(len(object));
+    for (int i = 0; i < len(object); ++i)
+      vars[i] = convertPythonObjectToVariable(object[i]);
+    return makeVariable(classname, vars);
+  }
+
   {
     DenseMatrixExtractor e(object);
-    if (e.check())
-    {
-      return makeDatatypeVariable(e);
-    }
+    if (e.check()) return makeDatatypeVariable(e);
   }
+
   {
     SparseRowMatrixExtractor e(object);
-    if (e.check())
-    {
-      return makeDatatypeVariable(e);
-    }
+    if (e.check()) return makeDatatypeVariable(e);
   }
```

Comment:
I was able to check the type without attempting an extraction on every type by checking for the python object's class name. Do the types DenseMatrix, SparseRowMatrix, and Field have expected class names of python objects I could use to clean up the remainder of this function?

Reply:
They're just particularly-laid-out lists and dicts, so checking the classname probably won't work, and might've broken matrices on line 515.

Reply:
If their storage also uses lists, we won't be able to distinguish between a matrix and a regular list of floats. I think we should type check the inputs of the set state function. Currently, we can set any parameter to any datatype. If we verify that the parameter's datatype matches the one that is stored, we can stop user error through the interface and decide which extractors to try between fields, matrices, and lists.

There is a slight problem with type checking being based on current state types. When you save a network and change the data type, you will only be able to set an object of the type that was used when creating the network. For example, a network made before this PR will not be able to set to CameraRotation and CameraLookAt to anything besides strings. Instead, we could verify that the type of data in the file matches the type of data being set as the default state. Here, we could attempt a conversion or just set to the default value.

Reply:
I ran into that same issue recently, thinking there should be some type-checking for Python state-setters. It would be annoying code to write, since you'd need to handle all the different type coercions that happen within core python, like converting ints to doubles. The Python interface philosophy has always been essentially user-driven, but user-beware. 

With these data checks looking at matrices and fields, those are usually used for data coming into the InterfaceWithPython module, rather than set_state calls. So the converters could be separated and made context dependent--use your new converter for state calls, and only check for matrices etc when data is being moved around. The catch is I think IWP uses transient_state variables to move the big datatypes around, so there's more special-case code to write. 

All-in-all, this is a good idea, but a lot more test code needs to be written.

Reply:
I'm not sure what you mean by making the converters context dependent. Also, what data is being moved around?